lib/trip_dispatcher.rb

```diff
@@ -3,39 +3,96 @@
 
 require_relative 'passenger'
 require_relative 'trip'
+require_relative 'driver'
 
 module RideShare
   class TripDispatcher
     attr_reader :drivers, :passengers, :trips
-
+    
     def initialize(directory: './support')
       @passengers = Passenger.load_all(directory: directory)
       @trips = Trip.load_all(directory: directory)
+      @drivers = Driver.load_all(directory: directory)
       connect_trips
     end
-
+    
     def find_passenger(id)
       Passenger.validate_id(id)
       return @passengers.find { |passenger| passenger.id == id }
     end
-
+    
+    def find_driver(id)
+      Driver.validate_id(id)
+      return @drivers.find { |driver| driver.id == id }
+    end
+    
     def inspect
       # Make puts output more useful
       return "#<#{self.class.name}:0x#{object_id.to_s(16)} \
-              #{trips.count} trips, \
-              #{drivers.count} drivers, \
-              #{passengers.count} passengers>"
+      #{trips.count} trips, \
+      #{drivers.count} drivers, \
+      #{passengers.count} passengers>"
     end
-
+    
+    def select_driver_by_history(available_driver_list)
+      # first select the driver who has never driven before
+      selected_driver = available_driver_list.find do |one_driver|
+        one_driver.trips.length == 0
+      end
+      return selected_driver if !(selected_driver.nil?)
+      
+      # then select the driver whose most recent trip ended the longest time ago
+      selected_driver = available_driver_list.max_by do |individual_driver|
+        (Time.new - individual_driver.trips.last.end_time)
+      end
+      return selected_driver
+    end
+    
+    def find_available_driver
+      all_available_drivers = []
+      drivers.each do |driver|
+        if driver.status == :AVAILABLE && driver.find_ongoing_trips.length == 0
+          all_available_drivers << driver   
+        end
+      end
+      if all_available_drivers.length == 0
+        raise ArgumentError, "No Available Drivers at this Time."
```

Comment:
This is not an ArgumentError because there are no bad arguments (`find_available_driver` doesn't take any arguments!). I would recommend using StandardError.

```suggestion
        raise StandardError, "No Available Drivers at this Time."
```